File: Sources/BraintreeCard/BTCardClient.m
#import "BTCardClient_Internal.h"
#import "BTCardNonce_Internal.h"
#import "BTCard_Internal.h"
#import "BTCardAnalytics.h"

// MARK: - Objective-C File Imports for Package Managers
#if __has_include(<Braintree/BraintreeCard.h>) // CocoaPods
#import <Braintree/BTCardRequest.h>

#elif SWIFT_PACKAGE // SPM
#import <BraintreeCard/BTCardRequest.h>

#else // Carthage
#import <BraintreeCard/BTCardRequest.h>

#endif

// MARK: - Swift File Imports for Package Managers
#if __has_include(<Braintree/Braintree-Swift.h>) // CocoaPods
#import <Braintree/Braintree-Swift.h>

#elif SWIFT_PACKAGE                              // SPM
/* Use @import for SPM support
 * See https://forums.swift.org/t/using-a-swift-package-in-a-mixed-swift-and-objective-c-project/27348
 */
@import BraintreeCore;

#elif __has_include("Braintree-Swift.h")         // CocoaPods for ReactNative
/* Use quoted style when importing Swift headers for ReactNative support
 * See https://github.com/braintree/braintree_ios/issues/671
 */
#import "Braintree-Swift.h"

#else                                            // Carthage
#import <BraintreeCore/BraintreeCore-Swift.h>
#endif

NSString *const BTCardClientErrorDomain = @"com.braintreepayments.BTCardClientErrorDomain";
NSString *const BTCardClientGraphQLTokenizeFeature = @"tokenize_credit_cards";

@interface BTCardClient ()
@end

@implementation BTCardClient

- (instancetype)initWithAPIClient:(BTAPIClient *)apiClient {
    if (!apiClient) {
        return nil;
    }
    if (self = [super init]) {
        self.apiClient = apiClient;
    }
    return self;
}

- (instancetype)init {
    return nil;
}

- (void)tokenizeCard:(BTCard *)card completion:(void (^)(BTCardNonce *tokenizedCard, NSError *error))completion {
    [self.apiClient sendAnalyticsEvent: BTCardAnalytics.cardTokenizeStarted];
    BTCardRequest *request = [[BTCardRequest alloc] initWithCard:card];

    if (!self.apiClient) {
        NSError *error = [NSError errorWithDomain:BTCardClientErrorDomain
                                             code:BTCardClientErrorTypeIntegration
                                         userInfo:@{NSLocalizedDescriptionKey: @"BTCardClient tokenization failed because BTAPIClient is nil."}];
        [self.apiClient sendAnalyticsEvent:BTCardAnalytics.cardTokenizeFailed];
        completion(nil, error);
        return;
    }

    [self.apiClient fetchOrReturnRemoteConfiguration:^(BTConfiguration * _Nullable configuration, NSError * _Nullable error) {
        if (error) {
            [self.apiClient sendAnalyticsEvent:BTCardAnalytics.cardTokenizeFailed];
            completion(nil, error);
            return;
        }

        if ([self isGraphQLEnabledForCardTokenization:configuration]) {
            
            if (request.card.authenticationInsightRequested && !request.card.merchantAccountID) {
                NSError *error = [NSError errorWithDomain:BTCardClientErrorDomain
                                                     code:BTCardClientErrorTypeIntegration
                                                 userInfo:@{NSLocalizedDescriptionKey: @"BTCardClient tokenization failed because a merchant account ID is required when authenticationInsightRequested is true."}];
                
                [self.apiClient sendAnalyticsEvent:BTCardAnalytics.cardTokenizeFailed];
                completion(nil, error);
                return;
            }
            
            NSDictionary *parameters = [request.card graphQLParameters];
            [self.apiClient POST:@""
                      parameters:parameters
                        httpType:BTAPIClientHTTPServiceGraphQLAPI
                      completion:^(BTJSON * _Nullable body, __unused NSHTTPURLResponse * _Nullable response, NSError * _Nullable error)
             {
                 if (error) {
                     if (error.code == BTCoreConstants.networkConnectionLostCode) {
                         [self.apiClient sendAnalyticsEvent:BTCardAnalytics.cardTokenizeNetworkConnectionLost];
                         [self.apiClient sendAnalyticsEvent:BTCardAnalytics.cardTokenizeFailed];
                         completion(nil, error);
                         return;
                     }
                     NSHTTPURLResponse *response = error.userInfo[BTCoreConstants.urlResponseKey];
                     NSError *callbackError = error;

                     if (response.statusCode == 422) {
                         if (error.userInfo) {
                             callbackError = [self constructCallbackErrorForErrorUserInfo:error.userInfo error:error];
                         }
                     }
                     
                     [self.apiClient sendAnalyticsEvent:BTCardAnalytics.cardTokenizeFailed];
                     completion(nil, callbackError);
                     return;
                 }

                BTJSON *cardJSON = body[@"data"][@"tokenizeCreditCard"];
                if (cardJSON.asError) {
                    [self.apiClient sendAnalyticsEvent:BTCardAnalytics.cardTokenizeFailed];
                    completion(nil, cardJSON.asError);
                    return;
                }

                BTCardNonce *cardNonce = [BTCardNonce cardNonceWithGraphQLJSON:cardJSON];
                [self.apiClient sendAnalyticsEvent:BTCardAnalytics.cardTokenizeSucceeded];
                completion(cardNonce, nil);
                return;
             }];
        } else {
            NSDictionary *parameters = [self clientAPIParametersForCard:request];
            [self.apiClient POST:@"v1/payment_methods/credit_cards"
                      parameters:parameters
                      completion:^(BTJSON *body, __unused NSHTTPURLResponse *response, NSError *error)
             {
                 if (error != nil) {
                     if (error.code == BTCoreConstants.networkConnectionLostCode) {
                         [self.apiClient sendAnalyticsEvent:BTCardAnalytics.cardTokenizeNetworkConnectionLost];
                         [self.apiClient sendAnalyticsEvent:BTCardAnalytics.cardTokenizeFailed];
                         completion(nil, error);
                         return;
                     }
                     NSHTTPURLResponse *response = error.userInfo[BTCoreConstants.urlResponseKey];
                     NSError *callbackError = error;

                     if (response.statusCode == 422) {
                         if (error.userInfo) {
                             callbackError = [self constructCallbackErrorForErrorUserInfo:error.userInfo error:error];
                         }
                     }
                     
                     [self.apiClient sendAnalyticsEvent:BTCardAnalytics.cardTokenizeFailed];
                     completion(nil, callbackError);
                     return;
                 }
                
                 BTJSON *cardJSON = body[@"creditCards"][0];
                 if (cardJSON.asError) {
                    [self.apiClient sendAnalyticsEvent:BTCardAnalytics.cardTokenizeFailed];
                    completion(nil, cardJSON.asError);
                    return;
                 }
                
                 BTCardNonce *cardNonce = [BTCardNonce cardNonceWithJSON:cardJSON];
                 [self.apiClient sendAnalyticsEvent:BTCardAnalytics.cardTokenizeSucceeded];
                 completion(cardNonce, nil);
                 return;
             }];
        }
    }];
}

#pragma mark - Helpers

+ (NSDictionary *)validationErrorUserInfo:(NSDictionary *)userInfo {
    NSMutableDictionary *mutableUserInfo = [userInfo mutableCopy];
    BTJSON *jsonResponse = userInfo[BTCoreConstants.jsonResponseBodyKey];
    if ([jsonResponse asDictionary]) {
        mutableUserInfo[@"BTCustomerInputBraintreeValidationErrorsKey"] = [jsonResponse asDictionary];

        NSString *errorMessage = [jsonResponse[@"error"][@"message"] asString];
        if (errorMessage) {
            mutableUserInfo[NSLocalizedDescriptionKey] = errorMessage;
        }

        BTJSON *fieldError = [jsonResponse[@"fieldErrors"] asArray].firstObject;
        BTJSON *firstFieldError = [fieldError[@"fieldErrors"] asArray].firstObject;
        NSString *firstFieldErrorMessage = [firstFieldError[@"message"] asString];
        if (firstFieldErrorMessage) {
            mutableUserInfo[NSLocalizedFailureReasonErrorKey] = firstFieldErrorMessage;
        }
    }
    return [mutableUserInfo copy];
}

- (NSDictionary *)clientAPIParametersForCard:(BTCardRequest *)request {
    NSMutableDictionary *parameters = [NSMutableDictionary new];
    if (request.card.parameters) {
        parameters[@"credit_card"] = [request.card.parameters copy];
    }
    parameters[@"_meta"] = @{
                             @"source" : self.apiClient.metadata.sourceString,
                             @"integration" : self.apiClient.metadata.integrationString,
                             @"sessionId" : self.apiClient.metadata.sessionID,
                             };
    
    if (request.card.authenticationInsightRequested) {
        parameters[@"authenticationInsight"] = @YES;
        parameters[@"merchantAccountId"] = request.card.merchantAccountID;
    }

    return [parameters copy];
}

- (BOOL)isGraphQLEnabledForCardTokenization:(BTConfiguration *)configuration {
    NSArray *graphQLFeatures = [configuration.json[@"graphQL"][@"features"] asStringArray];

    return graphQLFeatures && [graphQLFeatures containsObject:BTCardClientGraphQLTokenizeFeature];
}

- (NSError *)constructCallbackErrorForErrorUserInfo:(NSDictionary *)errorUserInfo error:(NSError *)error {
    NSError *callbackError = error;
    BTJSON *errorCode = nil;
    
    BTJSON *errorResponse = [error.userInfo objectForKey:BTCoreConstants.jsonResponseBodyKey];
    BTJSON *fieldErrors = [errorResponse[@"fieldErrors"] asArray].firstObject;
    errorCode = [fieldErrors[@"fieldErrors"] asArray].firstObject[@"code"];

    if (errorCode == nil) {
        BTJSON *errorResponse = [errorUserInfo objectForKey:BTCoreConstants.jsonResponseBodyKey];
        errorCode = [errorResponse[@"errors"] asArray].firstObject[@"extensions"][@"legacyCode"];
    }

    // Gateway error code for card already exists
    if ([errorCode.asString  isEqual: @"81724"]) {
        callbackError = [NSError errorWithDomain:BTCardClientErrorDomain
                                            code:BTCardClientErrorTypeCardAlreadyExists
                                        userInfo:[self.class validationErrorUserInfo:error.userInfo]];
    } else {
        callbackError = [NSError errorWithDomain:BTCardClientErrorDomain
                                            code:BTCardClientErrorTypeCustomerInputInvalid
                                        userInfo:[self.class validationErrorUserInfo:error.userInfo]];
    }
    return callbackError;
}

@end

File: Sources/BraintreeCard/BTCardClient.swift
import Foundation

#if canImport(BraintreeCore)
import BraintreeCore
#endif

/// Used to process cards
@objc public class BTCardClient: NSObject {

    // MARK: - Internal Properties

    /// Exposed for testing to get the instance of BTAPIClient
    var apiClient: BTAPIClient

    let graphQLTokenizeFeature: String = "tokenize_credit_cards"

    // MARK: - Initializer

    /// Creates a card client
    /// - Parameter authorization: A valid client token or tokenization key used to authorize API calls.
    @objc(initWithAuthorization:)
    public init(authorization: String) {
        self.apiClient = BTAPIClient(authorization: authorization)
    }

    // MARK: - Public Methods

    /// Tokenizes a card
    /// - Parameters:
    ///    - card: The card to tokenize.
    ///    - completion: A completion block that is invoked when card tokenization has completed. If tokenization succeeds,
    ///    `tokenize` will contain a nonce and `error` will be `nil`; if it fails, `tokenize` will be `nil` and `error`will describe the failure.
    @objc(tokenizeCard:completion:)
    public func tokenize(_ card: BTCard, completion: @escaping (BTCardNonce?, Error?) -> Void) {
        Task { @MainActor in
            do {
                let cardNonce = try await tokenize(card)
                completion(cardNonce, nil)
            } catch {
                completion(nil, error)
            }
        }
    }

    /// Tokenizes a card
    /// - Parameter card: The card to tokenize.
    /// - Returns: On success, you will receive an instance of `BTCardNonce`
    /// - Throws: An `Error` describing the failure
    public func tokenize(_ card: BTCard) async throws -> BTCardNonce {
        apiClient.sendAnalyticsEvent(BTCardAnalytics.cardTokenizeStarted)

        do {
            let configuration = try await apiClient.fetchOrReturnRemoteConfiguration()

            if isGraphQLEnabled(for: configuration) {
                if card.authenticationInsightRequested && card.merchantAccountID == nil {
                    throw BTCardError.integration
                }

                let parameters = card.graphQLParameters()

                do {
                    let (body, _) = try await apiClient.post("", parameters: parameters, httpType: .graphQLAPI)

                    let cardJSON: BTJSON = body?["data"]["tokenizeCreditCard"] ?? BTJSON()

                    if let cardJSONError = cardJSON.asError() {
                        notifyFailure(with: cardJSONError)
                        throw cardJSONError
                    }

                    let cardNonce = BTCardNonce(graphQLJSON: cardJSON)
                    notifySuccess()
                    return cardNonce
                } catch {
                    let nsError = error as NSError
                    let response: HTTPURLResponse? = nsError.userInfo[BTCoreConstants.urlResponseKey] as? HTTPURLResponse
                    var callbackError: Error = error

                    if response?.statusCode == 422 {
                        callbackError = constructCallbackError(with: nsError.userInfo, error: nsError) ?? error
                    }

                    notifyFailure(with: callbackError)
                    throw callbackError
                }
            } else {
                let parameters = card.parameters(apiClient: apiClient)

                do {
                    let (body, _) = try await apiClient.post("v1/payment_methods/credit_cards", parameters: parameters)

                    let cardJSON: BTJSON = body?["creditCards"][0] ?? BTJSON()

                    if let cardJSONError = cardJSON.asError() {
                        notifyFailure(with: cardJSONError)
                        throw cardJSONError
                    }

                    let cardNonce = BTCardNonce(json: cardJSON)
                    notifySuccess()
                    return cardNonce
                } catch {
                    let nsError = error as NSError
                    let response: HTTPURLResponse? = nsError.userInfo[BTCoreConstants.urlResponseKey] as? HTTPURLResponse
                    var callbackError: Error = error

                    if response?.statusCode == 422 {
                        callbackError = constructCallbackError(with: nsError.userInfo, error: nsError) ?? error
                    }

                    notifyFailure(with: callbackError)
                    throw callbackError
                }
            }
        } catch {
            notifyFailure(with: error)
            throw error
        }
    }

    // MARK: - Private Methods

    private func isGraphQLEnabled(for configuration: BTConfiguration) -> Bool {
        if let graphQLFeatures = configuration.json?["graphQL"]["features"].asStringArray() {
            return !graphQLFeatures.isEmpty && graphQLFeatures.contains(graphQLTokenizeFeature)
        }

        return false
    }

    // MARK: - Error Construction Methods

    /// Convenience helper method for creating friendlier, more human-readable userInfo dictionaries for 422 HTTP errors
    private func validationError(with userInfo: [String: Any]) -> [String: Any] {
        var finalUserInfo: [String: Any] = userInfo
        let jsonResponse: BTJSON? = userInfo[BTCoreConstants.jsonResponseBodyKey] as? BTJSON

        if let jsonDictionary = jsonResponse?.asDictionary() {
            finalUserInfo["BTCustomerInputBraintreeValidationErrorsKey"] = jsonDictionary
        }

        if let errorMessage = jsonResponse?["error"]["message"].asString() {
            finalUserInfo[NSLocalizedDescriptionKey] = errorMessage
        }

        let fieldError: BTJSON? = jsonResponse?["fieldErrors"].asArray()?.first
        let firstFieldError: BTJSON? = fieldError?["fieldErrors"].asArray()?.first

        if let firstFieldErrorMessage = firstFieldError?["message"].asString() {
            finalUserInfo[NSLocalizedFailureReasonErrorKey] = firstFieldErrorMessage
        }

        return finalUserInfo
    }

    private func constructCallbackError(with errorUserInfo: [String: Any]?, error: NSError?) -> Error? {
        let errorResponse: BTJSON? = error?.userInfo[BTCoreConstants.jsonResponseBodyKey] as? BTJSON
        let fieldErrors: BTJSON? = errorResponse?["fieldErrors"].asArray()?.first

        var errorCode: BTJSON? = fieldErrors?["fieldErrors"].asArray()?.first?["code"]
        var callbackError: Error? = error

        if errorCode == nil {
            let errorResponse: BTJSON? = errorUserInfo?[BTCoreConstants.jsonResponseBodyKey] as? BTJSON
            errorCode = errorResponse?["errors"].asArray()?.first?["extensions"]["legacyCode"]
        }

        // Gateway error code for card already exists
        if errorCode?.asString() == "81724" {
            callbackError = BTCardError.cardAlreadyExists(validationError(with: error?.userInfo ?? [:]))
        } else {
            callbackError = BTCardError.customerInputInvalid(validationError(with: error?.userInfo ?? [:]))
        }

        return callbackError
    }

    // MARK: - Analytics Helper Methods

    private func notifySuccess() {
        apiClient.sendAnalyticsEvent(BTCardAnalytics.cardTokenizeSucceeded)
    }

    private func notifyFailure(with error: Error) {
        apiClient.sendAnalyticsEvent(BTCardAnalytics.cardTokenizeFailed, errorDescription: error.localizedDescription)
    }
}
